### PR TITLE
drivers: spi: sam: Rework driver to support async transceive

### PR DESCRIPTION
**Pull request by pdgendt**
This PR reworks the SPI SAM driver to operate using interrupts. This allows to use it in slave mode and with async transceives.

### CODE REVIEW

**Comment by nandojve** [comment]
Hi @pdgendt ,

I'll look on this after #50966 got merged.

**Comment by pdgendt** [comment]
> I'll look on this after #50966 got merged.

I missed that one, this will require some additional work after it's merged 😃

**Comment by pdgendt** [comment]
> I'll look on this after #50966 got merged.

@nandojve The DMA version has been merged, however I would still like to get this PR in as well. Some questions:

- Should DMA + ASYNC work?
- I added RDRF interrupt in `spi_sam_configure`, does this conflict with DMA?
- I added CSAAT, which requires `SPI_CR_LASTXFER` to be set on the last byte, can this work with DMA?

Thanks

**Comment by nandojve** [comment]
Hi @pdgendt ,

> Should DMA + ASYNC work?

To offload CPU and consume less power answer is yes. The current DMA status is that DMA was just introduced and still not full supported. The driver still spinning wait transfer finish instead put CPU into LPM and wait an interrupt. No ASYNC support.
If I remember correctly you are using SAM4S variant and PDC is not current supported, only XDMAC from SAMS/E/V7x.
So, there is a road to implement all these features.

> I added RDRF interrupt in spi_sam_configure, does this conflict with DMA?

I never worked with slave mode on this devices. By what I understood, yes, you need that INT for slave mode but not necessarily when using DMA. I think RDRF status should trigger DMA when you are using it and in that sense RDRF int should be disabled. For instance, I guess you can use NSSR int to identify end transmission and since you know you system, you can have buffers allocated for the worst case.

> I added CSAAT, which requires SPI_CR_LASTXFER to be set on the last byte, can this work with DMA?

You need be careful with this option because driver should check cs_hold and lock_on operation bits. The other thing is the driver may use GPIOs to drive CS.

In summary, there are a lot of place to improve and a few can create issues. My suggestion would be to you improve driver first with interrupts, since you needed. After that you add the slave feature and work on other topics later.

**Review by tbursztyka — DISMISSED** [review]
Looks pretty fine to me.
Someone who knows Atmel's SAM would be nice for reviewing this PR

**Review by teburd — COMMENTED** [review]
One attribute that's quite nice with the old code is that for short transfers with fast SPI clocks it could do it without forcing context switches due to interrupts. The FIFO seems to be really small so its not clear using an interrupt driven mode is going to be super beneficial over using DMA in that case.

**Comment by pdgendt** [comment]
Hi @nandojve, I've updated the PR (finally).

Async support should be enabled now, I've also added `CSAAT` as a device tree binding to make this optional. Slave mode has been left out since I need to test this, it could be done in a new PR.

@teburd Could you verify that DMA isn't broken? I use SAM4S so I'm not able to verify this. Also see [my comment](https://github.com/zephyrproject-rtos/zephyr/pull/50966/commits/9d05dbd57b9c6137ff22695cc2ad1e6f1422d471#r1082385935) on your DMA PR which I think has a bug. I've fixed it in my interrupt rewrite.

**Comment by pdgendt** [comment]
Updated with DMA fix, @teburd can you test the updated version?

**Comment by nandojve** [comment]
Hi @pdgendt ,

Due to changes I'll need to test carefully and may take time.

One question about `csaat`, I'm not sure but maybe should it be defined on each transfer ? I mean, since controller may need operate with multiple devices and different behaviors.

**Comment by teburd** [comment]
> Updated with DMA fix, @teburd can you test the updated version?

I'll try and get to this sometime this week

**Comment by pdgendt** [comment]
> Due to changes I'll need to test carefully and may take time.

I guess it won't get in 3.3? That's fine for me.

> One question about `csaat`, I'm not sure but maybe should it be defined on each transfer ? I mean, since controller may need operate with multiple devices and different behaviors.

I've left out the `csaat` for now, it's indeed tied to the peripheral and not the controller itself. I will update and create a new PR after this gets merged.

**Comment by pdgendt** [comment]
Hi @teburd / @nandojve ,

I'd like to know what to do to get this PR moving.

> One attribute that's quite nice with the old code is that for short transfers with fast SPI clocks it could do it without forcing context switches due to interrupts. The FIFO seems to be really small so its not clear using an interrupt driven mode is going to be super beneficial over using DMA in that case.

Sure, but it doesn't work with the `ASYNC` API. Should I introduce some `kconfig` option to select either `SYNC+DMA` vs `ASYNC` or does that complicate the driver too much?

**Comment by teburd** [comment]
Not removing the polling transfer and making interrupt based transfers entirely opt-in would be ok with me.

**Comment by pdgendt** [comment]
> * make it entirely optional

I can try to refactor it this way.

> * change the IRQ_CONNECT to IRQ_DIRECT_CONNECT and avoid some of the overhead here for the very fast interrupt rate, perhaps only calling the zephyr pre/post ambles after N many bytes or at the end of the transfer. That way you sort of fake out a longer FIFO.

I'm not familiar with `IRQ_DIRECT_CONNECT`, I'll look into this.

> DMA should still probably be used when the transfers are longer than some number of bytes for sync or async as its beneficial in either case to free up the CPU time.

DMA isn't available on al boards that use this driver (my sam4s has no DMA).

**Comment by teburd** [comment]
> DMA isn't available on al boards that use this driver (my sam4s has no DMA).

Ahhhh understood, I get the desire to have the interrupt driven transfers more so now.
